server/src/main/java/io/crate/execution/engine/aggregation/impl/TopKAggregation.java

```diff
@@ -180,31 +185,120 @@ public State reduce(RamAccounting ramAccounting, State state1, State state2) {
+        }
+        return getDocValueAggregator(reference, (int) limit.value());
+    }
+
+    @Nullable
+    private DocValueAggregator<?> getDocValueAggregator(Reference ref, int limit) {
+        return switch (ref.valueType().id()) {
+            case LongType.ID -> new SortedNumericDocValueAggregator<>(
+                ref.storageIdent(),
+                (ramAccounting, _, _) -> topKState(ramAccounting, limit),
+                (values, state) -> {
+                    state.itemsSketch.update(values.nextValue());
+                }
+            );
+            case StringType.ID -> new BinaryDocValueAggregator<>(
+                ref.storageIdent(),
+                (ramAccounting, _, _) -> topKState(ramAccounting, limit),
+                (values, state) -> {
+                    long ord = values.nextOrd();
+                    BytesRef value = values.lookupOrd(ord);
+                    state.itemsSketch.update(value.utf8ToString());
+                }
+            );
+            default -> null;
+        };
+    }
+
     sealed interface State {
+
         State EMPTY = new Empty();
+
+        List<Map<String, Object>> result();
+    }
+
+    record Empty() implements State {
+
+        @Override
+        public List<Map<String, Object>> result() {
+            return List.of();
+        }
     }
 
-    record Empty() implements State { }
+    record TopKState(ItemsSketch<Object> itemsSketch, int limit) implements State {
+
+        static long SHALLOW_SIZE = RamUsageEstimator.shallowSizeOfInstance(TopKState.class);
+
+        public List<Map<String, Object>> result() {
+            if (itemsSketch.isEmpty()) {
+                return List.of();
+            }
+            ItemsSketch.Row<Object>[] frequentItems = itemsSketch.getFrequentItems(ErrorType.NO_FALSE_NEGATIVES);
+            int limit = Math.min(frequentItems.length, this.limit);
+            var result = new ArrayList<Map<String, Object>>(limit);
+            for (int i = 0; i < limit; i++) {
+                var item = frequentItems[i];
+                result.add(Map.of("item", item.getItem(), "frequency", item.getEstimate()));
+            }
+            return result;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            TopKState topKState = (TopKState) o;
+            return limit == topKState.limit && Objects.equals(result(), topKState.result());
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(result(), limit);
+        }
+    }
 
-    record TopKState(ItemsSketch<Object> itemsSketch, int limit) implements State { }
+    private static TopKState topKState(RamAccounting ramAccounting, int limit) {
+        int maxMapSize = maxMapSize(limit);
+        ramAccounting.addBytes(calculateRamUsage(maxMapSize));
+        ramAccounting.addBytes(TopKState.SHALLOW_SIZE);
```

Comment:
```suggestion
        ramAccounting.addBytes(calculateRamUsage(maxMapSize) + TopKState.SHALLOW_SIZE);
```

Depending on the type of `RamAccounting`, `addBytes` can have a bit of overhead